File: problem_solving/graph/py/547_number_of_provinces.py
'''
https://leetcode.com/problems/number-of-provinces/description/

this can be solved by bfs, dfs & disjoint sets.
'''

def number_of_provinces_dfs(connections: list[list[int]]) -> int:
    visited = set()
    rows = len(connections)
    cols = len(connections[0])
    
    def dfs(r):
        visited.add(r)
        for c in range(cols):
            if connections[r][c] != 1:
                continue
            if c in visited:
                continue
            dfs(c)   

    provinces = 0
    
    for r in range(len(connections)):
        if r in visited:
            continue
        provinces += 1
        dfs(r)

    return provinces


connections = [[1,1,0],[1,1,0],[0,0,1]]
#Output: 2
provinces = number_of_provinces_bfs(connections)
print("provinces: " + str(provinces))


connections = [[1,0,0],[0,1,0],[0,0,1]]
#Output: 3
provinces = number_of_provinces_bfs(connections)
print("provinces: " + str(provinces))

File: problem_solving/graph/cpp/547_number_of_provinces.cpp
/*
https://leetcode.com/problems/number-of-provinces/description/

this can be solved by bfs, dfs & disjoint sets.
*/

#include<iostream>
#include<tuple>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<unordered_map>
using namespace std;

class Solve {
    private:
    std::set<int> visited;
    std::unordered_map<int, std::vector<int>> provinces;

    int dfs(int src)  {
        visited.insert(src);
        auto neighbors = provinces[src];
        for(auto nei : neighbors) {
            if(visited.find(nei) == visited.end()) {
                dfs(nei);
            } 
        }
    }
    
    public:
    int number_of_provinces_dfs(std::vector<std::vector<int>> connections ) {
        visited.clear();
        provinces.clear();
        
        int result = 0;
        //create the adj. list.        
        for(int i = 0; i < connections.size(); ++i) {
            for(int j = 0; j < connections[0].size(); ++j) {
                if(connections[i][j] == 1) {
                    provinces[i].push_back(j);
                }
            }
        }
        
        //dfs into the vector if not already visited.
        for(int i = 0; i < connections.size(); ++i) {
            if(visited.find(i) != visited.end()) {
                continue;
            }
            result++;
            
            dfs(i);
            if(visited.size() == connections.size()) {
                break;
            }
        }
    
        return result;
    }
    
};


int main() {
    std::vector<std::vector<int>> connections = {{1,1,0},{1,1,0},{0,0,1}};
    //#Output: 2
    Solve solve;
    int provinces = solve.number_of_provinces_dfs(connections);
    std::cout << "provinces: " << provinces << std::endl;


    connections = {{1,0,0},{0,1,0},{0,0,1}};
    //    #Output: 3
    provinces = solve.number_of_provinces_dfs(connections);
    std::cout << "provinces: " << provinces << std::endl;
}